connector/src/main/java/org/geysermc/connector/utils/SettingsUtils.java

```diff
@@ -0,0 +1,162 @@
+            builder.addComponent(new LabelComponent(LanguageUtils.getPlayerLocaleString("geyser.settings.title.game_rules", language)));
+            for (GameRule gamerule : GameRule.values()) {
+                if (gamerule.equals(GameRule.UNKNOWN)) {
+                    continue;
+                }
+
+                if (Boolean.class.equals(gamerule.getType())) {
+                    builder.addComponent(new ToggleComponent(LocaleUtils.getLocaleString("gamerule." + gamerule.getJavaID(), language), GeyserConnector.getInstance().getWorldManager().getGameRuleBool(session, gamerule)));
+                } else if (Integer.class.equals(gamerule.getType())) {
+                    builder.addComponent(new InputComponent(LocaleUtils.getLocaleString("gamerule." + gamerule.getJavaID(), language), "", String.valueOf(GeyserConnector.getInstance().getWorldManager().getGameRuleInt(session, gamerule))));
+                }
+            }
+        }
+
+        session.setSettingsForm(builder.build());
+    }
+
+    /**
+     * Handle the settings form response
+     *
+     * @param session The session that sent the response
+     * @param response The response string to parse
+     * @return True if the form was parsed correctly, false if not
+     */
+    public static boolean handleSettingsForm(GeyserSession session, String response) {
+        CustomFormWindow settingsForm = session.getSettingsForm();
+        settingsForm.setResponse(response);
+
+        CustomFormResponse settingsResponse = (CustomFormResponse) settingsForm.getResponse();
+        int offset = 0;
+
+        offset++; // Client settings title
+
+        session.getWorldCache().setShowCoordinates(settingsResponse.getToggleResponses().get(offset));
+        offset++;
+
+        if (session.getOpPermissionLevel() >= 2 || session.hasPermission("geyser.settings.server")) {
+            offset++; // Server settings title
+
+            GameMode gameMode = GameMode.values()[settingsResponse.getDropdownResponses().get(offset).getElementID()];
+            if (gameMode != null && gameMode != session.getGameMode()) {
+                session.getConnector().getWorldManager().setPlayerGameMode(session, gameMode);
+            }
+            offset++;
+
+            Difficulty difficulty = Difficulty.values()[settingsResponse.getDropdownResponses().get(offset).getElementID()];
+            if (difficulty != null && difficulty != session.getWorldCache().getDifficulty()) {
+                session.getConnector().getWorldManager().setDifficulty(session, difficulty);
+            }
+            offset++;
+        }
+
+        if (session.getOpPermissionLevel() >= 2 || session.hasPermission("geyser.settings.gamerules")) {
+            offset++; // Game rule title
+
+            for (GameRule gamerule : GameRule.values()) {
+                if (gamerule.equals(GameRule.UNKNOWN)) {
+                    continue;
+                }
+
+                if (Boolean.class.equals(gamerule.getType())) {
```

Comment:
A small comment here would be nice - I get the gist but I did have to stare at it for a minute.